src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,96 @@
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqQueryToForEachProvider<TQueryExpression, TStatement> : CodeRefactoringProvider
+        where TQueryExpression : SyntaxNode
+        where TStatement : SyntaxNode
+    {
+        protected abstract string Title { get; }
+
+        protected abstract bool TryConvert(
+            TQueryExpression queryExpression,
+            SemanticModel semanticModel,
+            ISemanticFactsService semanticFacts,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdate);
+
+        protected abstract TQueryExpression FindNodeToRefactor(SyntaxNode root, TextSpan span);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var queryExpression = FindNodeToRefactor(root, context.Span);
+            if (queryExpression == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var semanticFacts = document.GetLanguageService<ISemanticFactsService>();
+            if (TryConvert(queryExpression, semanticModel, semanticFacts, cancellationToken, out DocumentUpdateInfo documentUpdateInfo))
+            {
+                context.RegisterRefactoring(
+                    new MyCodeAction(
+                        Title,
+                        c => Task.FromResult(document.WithSyntaxRoot(documentUpdateInfo.UpdateRoot(root)))));
+            }
+        }
+
+        /// <summary>
+        /// Handles information about updating the document with the refactoring.
+        /// </summary>
+        internal sealed class DocumentUpdateInfo
+        {
+            public readonly TStatement Source;
+            public readonly ImmutableArray<TStatement> Destinations;
+
+            public DocumentUpdateInfo(TStatement source, TStatement destination) : this(source, new[] { destination })
+            {
+            }
+
+            public DocumentUpdateInfo(TStatement source, IEnumerable<TStatement> destinations)
+            {
+                Source = source;
+                Destinations = ImmutableArray.CreateRange(destinations);
+            }
+
+            /// <summary>
+            /// Updates the root of the docuemtn with the document update.
+            /// </summary>
+            public SyntaxNode UpdateRoot(SyntaxNode root)
+            {
+                if (Destinations.Length == 1)
```

Comment:
>if (Destinations.Length == 1) [](start = 16, length = 29)

I am not familiar with ```ReplaceNode```. Could you elaborate what does this change make?

Reply:
There are two overloads: one accepts a collection of nodes and another a single node.
If we replace a node in statement of an  if-statement (without block), it cannot replace it with collection even if there is a just 1 element in it.

---
In reply to: [178423527](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423527) [](ancestors = 178423527)

Reply:
> If we replace a node in statement of an  if-statement (without block), it cannot replace it with collection even if there is a just 1 element in it.

Thanks for the explanation. It would be good to capture this in a comment 

---
In reply to: [178423607](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423607) [](ancestors = 178423607,178423527)

Reply:
added

---
In reply to: [178424286](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424286) [](ancestors = 178424286,178423607,178423527)